assembly/src/parser/grammar.lalrpop

```diff
@@ -0,0 +1,1378 @@
+use alloc::{
+    boxed::Box,
+    collections::{VecDeque, BTreeSet},
+    string::ToString,
+    sync::Arc,
+    vec::Vec,
+};
+use core::{marker::PhantomData, num::NonZeroU32};
+
+use smallvec::smallvec;
+use vm_core::{Felt, FieldElement, StarkField, crypto::hash::RpoDigest};
+
+use crate::{LibraryPath, LibraryNamespace, ProcedureName, ast::*, diagnostics::SourceFile};
+use super::{
+    HexEncodedValue, Token, ParseError, ParsingError,
+    LiteralErrorKind, HexErrorKind, Span, Spanned,
+    DocumentationType
+};
+
+grammar<'input>(
+    source_file: &Arc<SourceFile>,
+    interned: &mut BTreeSet<Arc<str>>,
+    _marker: PhantomData<&'input str>
+);
+
+// LEXER
+// ================================================================================================
+
+extern {
+    type Error = ParsingError;
+    type Location = u32;
+
+    enum Token<'input> {
+        bare_ident => Token::Ident(<&'input str>),
+        const_ident => Token::ConstantIdent(<&'input str>),
+        quoted_ident => Token::QuotedIdent(<&'input str>),
+        hex_value => Token::HexValue(<HexEncodedValue>),
+        doc_comment => Token::DocComment(<DocumentationType>),
+        uint => Token::Int(<u64>),
+        "add" => Token::Add,
+        "adv.insert_hdword" => Token::AdvInsertHdword,
+        "adv.insert_hperm" => Token::AdvInsertHperm,
+        "adv.insert_mem" => Token::AdvInsertMem,
+        "adv_loadw" => Token::AdvLoadw,
+        "adv_pipe" => Token::AdvPipe,
+        "adv_push" => Token::AdvPush,
+        "adv.push_ext2intt" => Token::AdvPushExt2intt,
+        "adv.push_mapval" => Token::AdvPushMapval,
+        "adv.push_mapvaln" => Token::AdvPushMapvaln,
+        "adv.push_mtnode" => Token::AdvPushMtnode,
+        "adv.push_sig" => Token::AdvPushSig,
+        "adv.push_smtpeek" => Token::AdvPushSmtpeek,
+        "adv.push_smtget" => Token::AdvPushSmtget,
+        "adv.push_smtset" => Token::AdvPushSmtset,
+        "adv.push_u64div" => Token::AdvPushU64Div,
```

Comment:
@bobbinth should we standardize the `adv_loadw`, `adv_pipe`, `adv_push` instructions to match the rest? (Or even better, drop `.` so that it is used to separate only arguments)

Reply:
I had the same thought, I was actually a bit surprised that those couple of instructions didn't follow the same convention, but didn't think it was appropriate to change that as part of this PR.

Reply:
The main reason why these are different is to differentiate between real instructions and decorators. So, anything which is `adv.xxx` is a decorator (i.e., it does not advance the clock cycle of the VM and does not affect the VM state outside of the advice provider). The `adv_xxx` are actual instructions which affect the state of the operand stack and/or memory.

Reply:
Is this distinction useful though? And if it is, is a single `.` the best way to express this?

Reply:
I think it is useful as something like `adv_push` has a very different semantic from `adv.push_u64div` - so, it is good to have syntactic differentiation. Whether using `.` is the best way to do that - I'm not sure. I picked it because conceptually in `adv.[action]` the `action` parametrizes what we should do to the advice provider (this is consistent with how we specify parameters for other instructions). But I'm open to other ways of doing this.

Reply:
I think my first impression upon seeing it is that I would assume they are both instructions, so we probably should differentiate decorators further somehow, such as prefixing them with `@` or something, e.g. `@adv.push_u64div` visually stands out as something special compared to regular instructions, and also makes it easy to spot them when skimming the code.

If we want to distinguish them somehow, I would suggest a prefix symbol from the set `@`, `!`, `%` or `$`. I'd be inclined to use `@` simply because it is used for attribute/decorator-ish things in other languages, but any of those symbols are free of any ambiguity issues in the current grammar.